### DISCUSSION TITLE
ControllerAdvice에 대해 설명해주세요

### DISCUSSION THREAD

**Discussion by nayonsoso** [2024-woowacourse-study/level-interview#137] [Spring MVC]
혹시나 나올 수 있는 꼬리 질문 : RestControllerAdvice와는 무슨 차이가 있나요?

**Comment by 3Juhwan** [comment]
`@ControllerAdvice`는 모든 컨트롤러에 대해 전역 기능을 제공하는 애너테이션입니다. `@ControllerAdvice`가 선언된 클래스에 `@ExceptionHandler`, `@InitBinder`, `@ModelAttribute`를 등록하면 예외 처리, 바인딩 등을 한 곳에서 처리할 수 있어, 코드의 중복을 줄이고 유지보수성을 높일 수 있습니다. `@ControllerAdvice`는 내부에 `@Component`가 포함되어 있어 컴포넌트 스캔 과정에서 빈으로 등록됩니다. `@RestControllerAdvice`는 내부에 `@ResponseBody`를 포함하여 `@ExceptionHandler`와 함께 사용될 때 예외 응답을 Json 형태로 내려준다는 특징이 있습니다.

**Comment by alstn113** [comment]
`@ControllerAdvice`는 `@ExceptionHandler`가 컨트롤러 코드에 들어가 정상 코드와 예외 처리 코드가 섞이는 것을 분리해준다. 또한 대상으로 지정한 여러 컨트롤러에 `@ExceptionHandler`, `@InitBinder`기능을 부여해주는 역할을 한다. `@ControllerAdvice`에 대상을 지정하지 않으면 모든 컨트롤러에 적용된다. 옵션으로 특정 애노테이션이 있는 컨트롤러를 지정할 수 있고, 특정 패키지를 직접 지정할 수도 있다. 특정 클래스를 지정할 수도 있다. 

`@RestControllerAdvice` 는 `@ControllerAdvice` 와 같고, `@ResponseBody` 가 추가되어 있다. `@Controller` , `@RestController` 의 차이와 같다.

참고: 김영한 MVC-2 9장 API 예외 처리
